src/adapters/IDB.js

```diff
@@ -1,22 +1,103 @@
 "use strict";
 
 import BaseAdapter from "./base.js";
-import { filterObject, omitKeys, sortObjects } from "../utils";
+import { filterObject, omitKeys, sortObjects, arrayEqual } from "../utils";
 
 const INDEXED_FIELDS = ["id", "_status", "last_modified"];
 
+/**
+ * Small helper that wraps the opening of an IndexedDB into a Promise.
+ *
+ * @param dbname          {String}   The database name.
+ * @param version         {Integer}  Schema version
+ * @param onupgradeneeded {Function} The callback to execute if schema is
+ *                                   missing or different.
+ * @return {Promise<IDBDatabase>}
+ */
+export async function open(dbname, { version, onupgradeneeded }) {
+  return new Promise((resolve, reject) => {
+    const request = indexedDB.open(dbname, version);
+    request.onupgradeneeded = event => {
+      const db = event.target.result;
+      db.onerror = event => reject(event.target.error);
+      return onupgradeneeded(event);
+    };
+    request.onerror = event => {
+      reject(event.target.error);
+    };
+    request.onsuccess = event => {
+      const db = event.target.result;
+      resolve(db);
+    };
+  });
+}
+
+/**
+ * Helper to run the specified callback in a single transaction on the
+ * specified store.
+ * The helper focuses on transaction wrapping into a promise.
+ *
+ * @param db           {IDBDatabase} The database instance.
+ * @param name         {String}      The store name.
+ * @param callback     {Function}    The piece of code to execute in the transaction.
+ * @param options      {Object}      Options.
+ * @param options.mode {String}      Transaction mode (default: read).
+ * @return {Promise} any value returned by the callback.
+ */
+export async function execute(db, name, callback, options = {}) {
+  const { mode } = options;
+  return new Promise((resolve, reject) => {
+    // On Safari, calling IDBDatabase.transaction with mode == undefined raises
+    // a TypeError.
+    const transaction = mode
+      ? db.transaction([name], mode)
+      : db.transaction([name]);
+    const store = transaction.objectStore(name);
+
+    // Let the callback abort this transaction.
+    const abort = e => {
+      transaction.abort();
+      reject(e);
+    };
+    // Execute the specified callback **synchronously**.
+    let result;
+    try {
+      result = callback(store, abort);
+    } catch (e) {
+      abort(e);
+    }
+    transaction.onerror = event => reject(event.target.error);
+    transaction.oncomplete = event => resolve(result);
+  });
+}
+
+/**
+ * Helper to wrap the deletion of an IndexedDB database into a promise.
+ *
+ * @param dbName {String} the database to delete
+ * @return {Promise}
+ */
+async function deleteDatabase(dbName) {
+  return new Promise((resolve, reject) => {
+    const request = indexedDB.deleteDatabase(dbName);
+    request.onsuccess = resolve();
```

Comment:
Oh. Very good catch.

Reply:
Does that mean there wasn't correct automated testing to catch that?

Reply:
Well, the deletion actually works and is tested. But this code could have cause race conditions issues since the promise would potentially resolve before the deletion occurs.